File: matlab/pulserationaldelay.m
function signalOut = pulserationaldelay(signalIn,Parms)
%PULSERATIONALDELAY applies rational transfer function with delay
%
% Syntax:	signalOut = pulsepolyratiodelay(signalIn,Parms)
% 
% Description:
% 
% pulserationaldelay calculates time-domain output with a rational transfer
% function together with a delay:
%
%           b(1) + b(2)*s + b(3)*s^2 + ... + b(nb)*s^(nb-1)
%   t =    ------------------------------------------------- exp(-s*eta),
%           a(1) + a(2)*s + a(3)*s^2 + ... + a(na)*s^(na-1)
% 
%   where s = i*w and w is the scaled angular frequency,
%   ranging over (-pi,pi). If Parms is a structure array, the transfer
%   function is a sum of rational functions, in which each term in the sum
%   is formed from the parameters contained in an array element.
%
% Inputs:
%   signalIn	input data vector, arbitrary units
%   Parms       Structure array of parameters:
%       .a          Denominator coefficients
%       .b          Numerator coefficients
%       .eta        Delay
%    
% 
% Outputs:
%   signalOut	output data vector, arbitrary units
% 
% 
% Other m-files required: none
% Subfunctions: none
% MAT-files required: none
%

% Author: J. Steven Dodge, 2005
% Physics Dept., Simon Fraser University
% Website: http://www.physics.sfu.ca/people/profiles/jsdodge
%
% $Id$

%------------- BEGIN CODE --------------

[N,c] = size(signalIn);
Nmax = floor((N-1)/2);
nTerm = length(Parms);

if c~=1
    warning('Multiple columns in data arrays, operating on first')
    signalIn=signalIn(:,1);
end

f = (1:Nmax)/N;
w = 2*pi*f;
w = w(:);
s = 1i*w;

% Fourier transform signal
signalInFt = fft(signalIn);
in = signalInFt(2:Nmax + 1);

% Initialize transfer function array
tFun = zeros(Nmax,1);

% Sum over terms
for iTerm = 1:nTerm

    na=length(Parms(iTerm).a)-1;
    nb=length(Parms(iTerm).b)-1;

    % Form the numerator and denominator
    nab = max(na,nb);
    kVal = (0:nab);
    vanderS = (s(:,ones(1,nab+1)).^kVal(ones(Nmax,1),:));

    denominator = vanderS(:,1:na+1)*Parms(iTerm).a;
    numerator   = vanderS(:,1:nb+1)*Parms(iTerm).b;

    tFun = tFun + (numerator./denominator).*exp(-Parms(iTerm).eta*s);
end
% Run through rational transfer function
out = tFun.*in;

if (rem(N,2)==0)
    signalOut = ...
        real(ifft([0;out;0;conj(flipud(out))]));
else
    signalOut = ...
        real(ifft([0;out;conj(flipud(out))]));
end
%------------- END OF CODE --------------

File: thztoolsPY/pulserationaldelay.py
import math
import numpy as np


def pulserelationdelay(signalin, parms):
    n = len(signalin)
    nmax = math.floor((n - 1) / 2)
    # nTerm = len(parms[0][0])

    w = np.ones(nmax)

    for i in range(len(w)):
        f = (i + 1) / n
        w[i] = 2 * np.pi * f
    s = 1j*w

    # Fourier transform signal
    signalinft = np.fft.fft(signalin)
    iin = signalinft[2:nmax + 1]

    for i in range(len(parms[0])):
        na = len(parms[0][i])
    for j in range(len(parms[0])):
        nb = len(parms[1][j])

    nab = max(na, nb)

    kval = np.zeros(nab)
    for k in range(len(kval)):
        kval[k] = k

    vanders = np.power(np.tile((np.reshape(w, (len(w), 1))), nab),
                       kval)  # first re shape w, expands it and then exponentiation

    denominator = vanders @ parms[0][0]
    numerator = vanders @ parms[1][0]

    tfun = np.zeros(nmax)

    for m in range(len(parms[2])):
        tfun = tfun + (numerator / denominator) * np.exp(parms[2][m] * s)

    out = tfun * iin

    if n % 2 == 0:
        signalout = np.real(np.fft.ifft(np.concatenate((np.array([0]), out, np.array([0]), np.conj(np.flipud(out))))))

    else:
        signalout = np.real(np.fft.ifft(np.concatenate((np.array([0]), out, np.conj(np.flipud(out))))))

    return signalout
